app/assets/javascripts/app/utils.js

```diff
@@ -11,18 +11,39 @@ $(document).on('click', dismiss, (e) => { $(e.target).parent().remove(); });
 // "page:load" event needed because of turbolinks, which overrides normal loading process
 $(document).on('ready page:load', () => {
   const message = '<div class="error-notify alert-danger p1 mt1 mb2">Please fill in all required' +
-    ' fields.</div>';
+    ' fields in their proper format.</div>';
 
   $('form').on('submit', function(e) {
     const $form = $(this);
-    const $fields = $form.find('[required]').filter(function() { return this.value === ''; });
+    const $required = $form.find('[required]').filter(function() { return this.value === ''; });
+    const $pattern = $form.find('[pattern]').filter(function() { return this.value; });
 
-    if ($fields.length) {
+    $('input', this).removeClass('border-red');
+    $('.error-notify').remove();
+
+    // Highlight any required fields without a value
+    if ($required.length) {
       e.preventDefault();
-      if (!$('.error-notify').length) {
-        $form.prepend(message);
-      }
-      $fields.each(function() { $(this).addClass('border-red'); });
+      $required.each(function() { $(this).addClass('border-red'); });
+    }
+
+    // Highlight any fields that do not match their defined pattern
+    let patternValid = true;
+    if ($pattern.length) {
+      $pattern.each(function() {
+        const $patternValue = $(this).prop('pattern');
+        const $regex = new RegExp(`^${$patternValue}$`);
+        if (!$(this).val().match($regex)) {
+          e.preventDefault();
+          $(this).addClass('border-red');
+          patternValid = false;
+        }
+      });
+    }
+
+    // Display error message if there are either missing fields or fields not in the proper format
+    if ($required.length || !patternValid) {
+      $form.prepend(message);
```

Comment:
How does the fallback mechanism work here?  IIUC your comment suggests that this code should only run on browsers that don't support html5?  I don't see how it conditionally executed.